### PR TITLE
I123 glo decoder error detector

### PR DESCRIPTION
**Pull request by dt-exafore**
The commit adds error detection algorithm for GLO nav messages + unit tests.
Requires: https://github.com/swift-nav/libswiftnav/pull/329

### CODE REVIEW

**Comment by dt-exafore** [comment]
@adel-mamin, please review.

File: src/nav_msg_glo.c [hunk 1]
```diff
@@ -33,14 +49,87 @@ void nav_msg_init_glo(nav_msg_glo_t *n)
...
+  p0 = parity(extract_word_glo(n, 1, 8));
+  c_sum = p0 ^ p1 ^ p2 ^ p3;
+
+  /* Now check C word to figure out is the string good, bad or
+   * correction is needed */
+  if ((!c_sum && !bit_set) || (1 == bit_set && c_sum)) /* case a) from ICD */
+
+    return 0; /* The string is good */
+
+  else if ((bit_set > 0 && !c_sum)
```
Comment (adel-mamin) [inline src/nav_msg_glo.c]: Should we remove 'else' here and below, since otherwise we always return?

File: src/nav_msg_glo.c [hunk 2]
```diff
@@ -33,14 +49,87 @@ void nav_msg_init_glo(nav_msg_glo_t *n)
...
 }
 
+/** The function performs data verification and error correction
+ * in received GLO navigation string. Refer to GLO ICD, section 4.7
+ * \param n pointer to GLO nav message structure
+ * \return -1 -- received string is bad and should be dropped out,
+ *          0 -- received string is good
+ *          >0 -- number of bit in n->string_bits to be corrected (inverted)
+ *                range[9..85]*/
+s8 error_detection_glo(nav_msg_glo_t *n)
```
Comment (adel-mamin) [inline src/nav_msg_glo.c]: Can 'n' be declared as 'const nav_msg_glo_t *n' in this function?

File: src/nav_msg_glo.c [hunk 3]
```diff
@@ -33,14 +49,87 @@ void nav_msg_init_glo(nav_msg_glo_t *n)
...
+  else if ((bit_set > 0 && !c_sum)
+           || (0 == bit_set && c_sum)) /* case c) from ICD */
+
+    return -1; /* multiple errors, bad string */
+
+  else if (bit_set > 1 && c_sum) { /* case b) from ICD */
+
+    u8 i_corr = (c & 0x7f) + 8 - k; /* define number of bit to be corrected */
+
+    if (i_corr > 85)
```
Comment (adel-mamin) [inline src/nav_msg_glo.c]: Is it worth adding a define for the constant '85'?

**Comment by adel-mamin** [comment]
Looks good to me.

File: src/nav_msg_glo.c [hunk 4]
```diff
@@ -33,17 +49,90 @@ void nav_msg_init_glo(nav_msg_glo_t *n)
...
  * \param bit_index number of bit the extract process start with. Range [1..85]
  * \param n_bits how many bits should be extracted [1..32]
  * \return word extracted from navigation string
  */
-u32 extract_word_glo(const nav_msg_glo_t *n, u16 bit_index, u8 n_bits)
+static u32 extract_word_glo(const nav_msg_glo_t *n, u16 bit_index, u8 n_bits)
 {
   assert(n_bits <= 32 && n_bits > 0);
-  assert(bit_index <= 85 && bit_index > 0);
+  assert(bit_index <= GLO_STR_LEN && bit_index > 0);
```
Comment (valeri-atamaniouk) [inline src/nav_msg_glo.c]: Return error if parameters are invalid. Do not assert here.

File: src/nav_msg_glo.c [hunk 5]
```diff
@@ -15,6 +15,13 @@
 #include <math.h>
 #include <libswiftnav/nav_msg_glo.h>
 #include <libswiftnav/time.h>
+#include <libswiftnav/logging.h>
+#include <libswiftnav/bits.h>
+
+#ifdef UNIT_TEST
```
Comment (gsmcmullin) [inline src/nav_msg_glo.c]: I'd prefer to move this to check_glo_decoder.c which includes this file.
```
#include <libswiftnav/logging.h>
#undef log_error
#define log_error(...)
#include "nav_msg_glo.c"
```

File: src/nav_msg_glo.c (hunk 1 above)
Comment (gsmcmullin) [inline src/nav_msg_glo.c]: Yes.  I'd prefer the else's removed, but it's not a big deal.

**Comment by gsmcmullin** [comment]
See my inline comments, but otherwise looks good.  I haven't checked the algorithm, I'll take that on faith.

File: src/nav_msg_glo.c (hunk 1 above)
Comment (dt-exafore) [inline src/nav_msg_glo.c]: Guys, it's already removed :)

**Comment by dt-exafore** [comment]
Corrected. Thanks @gsmcmullin, @adel-mamin and @valeri-atamaniouk

**Pull request merged**